src/pk/ecc/ecc_recover_key.c

```diff
@@ -0,0 +1,238 @@
+   @param hashlen     The length of the hash (octets)
+   @param recid       The recovery ID ("v"), can be -1 if signature contains it
+   @param sigformat   The format of the signature (ecc_signature_type)
+   @param key         The recovered public ECC key
+   @return CRYPT_OK if successful (even if the signature is not valid)
+*/
+int ecc_recover_key(const unsigned char *sig,  unsigned long siglen,
+                    const unsigned char *hash, unsigned long hashlen,
+                    int recid, int sigformat, ecc_key *key)
+{
+   ecc_point     *mG = NULL, *mQ = NULL, *mR = NULL;
+   void          *p, *m, *a, *b;
+   void          *r, *s, *v, *w, *t1, *t2, *u1, *u2, *v1, *v2, *e, *x, *y, *a_plus3;
+   void          *mu = NULL, *ma = NULL;
+   void          *mp = NULL;
+   int           err;
+   unsigned long pbits, pbytes, i, shift_right;
+   unsigned char ch, buf[MAXBLOCKSIZE];
+
+   LTC_ARGCHK(sig  != NULL);
+   LTC_ARGCHK(hash != NULL);
+   LTC_ARGCHK(key  != NULL);
+
+   /* BEWARE: requires sqrtmod_prime */
+   if (ltc_mp.sqrtmod_prime == NULL) {
+      return CRYPT_ERROR;
+   }
+
+   /* allocate ints */
+   if ((err = mp_init_multi(&r, &s, &v, &w, &t1, &t2, &u1, &u2, &v1, &v2, &e, &x, &y, &a_plus3, NULL)) != CRYPT_OK) {
+      return err;
+   }
+
+   p = key->dp.order;
+   m = key->dp.prime;
+   a = key->dp.A;
+   b = key->dp.B;
+   if ((err = mp_add_d(a, 3, a_plus3)) != CRYPT_OK) {
+      goto error;
+   }
+
+   /* allocate points */
+   mG = ltc_ecc_new_point();
+   mQ = ltc_ecc_new_point();
+   mR = ltc_ecc_new_point();
+   if (mR == NULL || mQ  == NULL || mG == NULL) {
+      err = CRYPT_MEM;
+      goto error;
+   }
+
+   if (sigformat == LTC_ECCSIG_ANSIX962) {
+      /* ANSI X9.62 format - ASN.1 encoded SEQUENCE{ INTEGER(r), INTEGER(s) }  */
+      if ((err = der_decode_sequence_multi_ex(sig, siglen, LTC_DER_SEQ_SEQUENCE | LTC_DER_SEQ_STRICT,
+                                     LTC_ASN1_INTEGER, 1UL, r,
+                                     LTC_ASN1_INTEGER, 1UL, s,
+                                     LTC_ASN1_EOL, 0UL, NULL)) != CRYPT_OK)                             { goto error; }
+   }
+   else if (sigformat == LTC_ECCSIG_RFC7518) {
+      /* RFC7518 format - raw (r,s) */
+      if ((siglen % 2) == 1) {
+         err = CRYPT_INVALID_PACKET;
+         goto error;
+      }
+      i = siglen / 2;
+      if ((err = mp_read_unsigned_bin(r, (unsigned char *)sig,   i)) != CRYPT_OK)                       { goto error; }
+      if ((err = mp_read_unsigned_bin(s, (unsigned char *)sig+i, i)) != CRYPT_OK)                       { goto error; }
+   }
+   else if (sigformat == LTC_ECCSIG_ETH) {
+      /* Bitcoin/Ethereum (v,r,s) format */
+      if (siglen != 65) { /* Only secp256k1 curves used */
+         err = CRYPT_INVALID_PACKET;
+         goto error;
+      }
+      i = (unsigned long)sig[64];
+      if ((i>=27) && (i<31)) i -= 27; /* Bitcoin/Ethereum add 27 to recovery ID to make it look less arbitrary */
```

Comment:
According to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md it looks like the value `27` is not that fixed as it looks; IIUC it can be `27`, `28`, `37`, `38` (and maybe more in the future). But I have not got deep enough into ETH yet.

Reply:
The original format is 27/28, and that's still valid even on the Ethereum main chain.

As you say, EIP-155 adds 35/36 _**plus 2x the chain ID**_ - which we cannot possibly know. That's why the library implementations tend to go with 0/1 and leave it up to the application to sort out. It's also not possible to represent 1337x2+35 (private chains) in a byte, so this format cannot be used for that.

29/30 are also possible on the original scheme, though the Yellow Paper incorrectly describes them as "infinite". However, most ETH implementations can't handle them, and since they only have about a 1-in-2^127 chance of occurring, nobody will ever see them in the life of the universe.

Basically, this entire space is an undocumented ill-conceived mess...

Reply:
OK, I suggest being more explicit and rename `LTC_ECCSIG_ETH` to `LTC_ECCSIG_ETH27` so that it is clear that it means *"Ethereum +27 El Clásico"* variant. The caller of `ecc_sign_hash_ex` gets the true recovery id in `*recid` anyway so they can handle it afterwards as they wish.

Reply:
OK, makes sense. On that basis, the main change will be in `ecc_sign_hash_ex` - we should probably support both 0/1 and 27/28 equivalently in `ecc_recover_key`, which is what that line does.

Reply:
Done. I've updated the docs and tests, removed any mention of Bitcoin (since I can't see any clear reference to non-DER signatures there), and also fixed a zero-padding bug when r or s are smaller than 2^251.